examples/erc1155/lib.rs

```diff
@@ -0,0 +1,738 @@
+    ///
+    /// The number of `values` specified to be transfer must match the number of `token_ids`,
+    /// otherwise this call will revert.
+    ///
+    /// Note that the call does not have to originate from the `from` account, and may originate
+    /// from any account which is approved to transfer `from`'s tokens.
+    #[ink(message)]
+    fn safe_batch_transfer_from(
+        &mut self,
+        from: AccountId,
+        to: AccountId,
+        token_ids: Vec<TokenId>,
+        values: Vec<Balance>,
+        data: Vec<u8>,
+    );
+
+    /// Query the balance of a specific token for the provided account.
+    #[ink(message)]
+    fn balance_of(&self, owner: AccountId, token_id: TokenId) -> Balance;
+
+    /// Query the balances for a set of tokens for a set of accounts.
+    ///
+    /// E.g use this call if you want to query what Alice and Bob's balances are for Tokens ID1 and
+    /// ID2.
+    ///
+    /// This will return all the balances for a given owner before moving on to the next owner. In
+    /// the example above this means that the return value should look like:
+    ///
+    /// [Alice Balance of Token ID1, Alice Balance of Token ID2, Bob Balance of Token ID2, Bob Balance of Token ID2]
+    #[ink(message)]
+    fn balance_of_batch(
+        &self,
+        owners: Vec<AccountId>,
+        token_ids: Vec<TokenId>,
+    ) -> Vec<Balance>;
+
+    /// Enable or disable a third party, known as an `operator`, to control all tokens on behalf of
+    /// the caller.
+    #[ink(message)]
+    fn set_approval_for_all(&mut self, operator: AccountId, approved: bool);
+
+    /// Query if the given `operator` is allowed to control all of `owner`'s tokens.
+    #[ink(message)]
+    fn is_approved_for_all(&self, owner: AccountId, operator: AccountId) -> bool;
+}
+
+/// The interface for an ERC-1155 Token Receiver contract.
+///
+/// The interface is defined here: `<https://eips.ethereum.org/EIPS/eip-1155>`.
+///
+/// Smart contracts which want to accept token transfers must implement this interface. By default
+/// if a contract does not support this interface any transactions originating from an ERC-1155
+/// compliant contract which attempt to transfer tokens directly to the contract's address must be
+/// reverted.
+#[ink::trait_definition]
+pub trait Erc1155TokenReceiver {
+    /// Handle the receipt of a single ERC-1155 token.
+    ///
+    /// This should be called by a compliant ERC-1155 contract if the intended recipient is a smart
+    /// contract.
+    ///
+    /// If the smart contract implementing this interface accepts token transfers then it must
+    /// return `ON_ERC_1155_RECEIVED_SELECTOR` from this function. To reject a transfer it must revert.
+    ///
+    /// Any callers must revert if they receive anything other than `ON_ERC_1155_RECEIVED_SELECTOR` as a return
+    /// value.
+    #[ink(message)]
+    fn on_erc_1155_received(
```

Comment:
Maybe it is the standard naming but I think it is pretty ugly to repeat yourself with `erc_1155` here.

Reply:
It's a little awkward, but I understand the naming choice. 

This is a way to handle incoming `ERC-1155` tokens and can be implemented by contracts which themselves are **not** `ERC-1155` contracts.

Reply:
Hmmm, I see, however in Rust we can also opt-in to have disambiguation by using `Erc1155::on_received` to state that this `on_received` is coming from the `Erc1155` trait. So we usually don't do this in the context of Rust and therefore I'd assume that an ink! example would use Rusty identifiers.

Reply:
That's a good point. I'll make it more Rusty.

FYI, these methods come from the `Erc1155TokenReceiver` trait, not the `Erc1155` trait. So the naming is even more awkward in that sense (`Erc1155TokenReceiver::on_erc_1155_received`)